No entity match error

Hi,

i get an error in protocoll.

[13:16:44] INFO: Starting Alexa Shopping List Sync
[2024-09-03 13:16:59 +0200] [85] [INFO] Running on http://0.0.0.0:5000 (CTRL + C to quit)
[2024-09-03 13:16:59,279] [ERROR] An error occurred while fetching items: Failed to get items: {'code': 'home_assistant_error', 'message': 'Service call requested response data but did not match any entities'}

Verison is 0.1.4, installed on Intel Nuc, running HASS OS.
There is one item in my Alexa-List, which I created new.

--- a/.github/ISSUE_TEMPLATE/language_support.md
+++ b/.github/ISSUE_TEMPLATE/language_support.md
@@ -1,0 +1,79 @@
+name: Language Support Request
+description: Request support for a new language in this project
+title: '[Language]: Add support for [my language]'
+labels:
+  - language
+  - enhancement
+body:
+  - type: dropdown
+    id: language
+    attributes:
+      label: Home Assistant Language
+      description: Language of your HA instance, set in Settings => System => General
+      options:
+        - Afrikaans
+        - العربية
+        - Български
+        - বাংলা
+        - Català
+        - Čeština
+        - Dansk
+        - Deutsch
+        - Schwyzerdütsch
+        - Ελληνικά
+        - English
+        - Español
+        - Eesti
+        - Euskara
+        - فارسی
+        - Suomi
+        - Français
+        - Français (Canada)
+        - Galician
+        - Gujarati
+        - עברית
+        - हिन्दी
+        - Hrvatski
+        - Magyar
+        - Bahasa Indonesia
+        - íslenska
+        - Italiano
+        - ქართული
+        - ಕನ್ನಡ
+        - 한국어
+        - Lëtzebuergesch
+        - Lietuvių
+        - Latviešu
+        - മലയാളം
+        - монгол хэл
+        - Bahasa Melayu
+        - Norsk Bokmål
+        - Nederlands
+        - Polski
+        - Português
+        - Português (Brasil)
+        - Română
+        - Русский
+        - Slovenčina
+        - Slovenščina
+        - Српски
+        - Svenska
+        - Kiswahili
+        - తెలుగు
+        - Türkçe
+        - Українська
+        - اُردُو
+        - Tiếng Việt
+        - 简体中文
+        - 繁體中文（香港）
+        - 繁體中文（台灣
+    validations:
+      required: true
+  - type: input
+    id: entity_id
+    attributes:
+      label: Entity ID
+      description: ' You can find the entity ID by going to `Developer Tools` => `States` and searching for `todo`'
+      placeholder: todo.einkaufsliste
+    validations:
+      required: true

--- a/alexa_shopping_list_sync/README.md
+++ b/alexa_shopping_list_sync/README.md
@@ -144,5 +144,15 @@ rest_command:
       Content-Type: application/json
 ```
 
+# Troubleshooting
+
+If you receive the following error message:
+
+```
+[ERROR] An error occurred while fetching items: Failed to get items: {'code': 'home_assistant_error', 'message': 'Service call requested response data but did not match any entities'}
+```
+
+This is because the add-on is unable to find the `todo.shopping_list` entity in Home Assistant. Please create an [issue](https://github.com/SavageCore/HA_Addons/issues/new?template=language_support.md) listing your Home Assistant language and the entity ID of your shopping list so I can add support for it. You can find the entity ID by going to `Developer Tools` => `States` and searching for `todo`. For example, Deutsch (German) is `todo.einkaufsliste`.
+
 [ko-fi-badge]: https://ko-fi.com/img/githubbutton_sm.svg
 [ko-fi-url]: https://ko-fi.com/E1E6P7VIQ

--- a/alexa_shopping_list_sync/app/classes/homeassistant_api.py
+++ b/alexa_shopping_list_sync/app/classes/homeassistant_api.py
@@ -1,0 +1,23 @@
+import requests
+
+
+class HomeAssistantAPI:
+
+    def __init__(self, url, token, logger=None):
+        if url.startswith("wss"):
+            self.url = url.replace("wss", "https", 1)
+        elif url.startswith("ws"):
+            self.url = url.replace("ws", "http", 1)
+        self.url = self.url.replace("websocket", "config", 1)
+        self.token = token
+        self.logger = logger
+
+    def get_current_language(self):
+        headers = {
+            "Authorization": f"Bearer {self.token}",
+            "Content-Type": "application/json",
+        }
+        response = requests.get(f"{self.url}", headers=headers)
+        response.raise_for_status()  # Raise an exception for HTTP errors
+        config_data = response.json()
+        return config_data.get("language", "en-GB")  # Default to 'en' if not found

--- a/alexa_shopping_list_sync/app/classes/homeassistant_websocket.py
+++ b/alexa_shopping_list_sync/app/classes/homeassistant_websocket.py
@@ -3,6 +3,8 @@ import json
 import logging
 
 import websockets
+from classes.homeassistant_api import HomeAssistantAPI
+from constants.entity_id_mapping import entity_id_mapping
 
 
 class HomeAssistantWebsocket:
@@ -17,6 +19,8 @@ class HomeAssistantWebsocket:
         self.logger = logger or logging.getLogger(
             __name__
         )  # Use provided logger or create one
+        self.ha_api = HomeAssistantAPI(url, token, logger=logger)
+        self.entity_id = self.get_entity_id_for_language()
 
     async def connect(self):
         """Establish a connection to the Home Assistant WebSocket API."""
@@ -69,6 +73,11 @@ class HomeAssistantWebsocket:
         if auth_response.get("type") != "auth_ok":
             raise Exception("Authentication failed")
 
+    def get_entity_id_for_language(self):
+        language = self.ha_api.get_current_language()
+
+        return entity_id_mapping.get(language, "todo.shopping_list")
+
     async def get_todo_list_items(self):
         """Fetch the current todo/shopping list items from Home Assistant."""
         if self.websocket is None or self.websocket.closed:
@@ -80,7 +89,7 @@ class HomeAssistantWebsocket:
             "domain": "todo",
             "service": "get_items",
             "return_response": True,
-            "service_data": {"entity_id": "todo.shopping_list"},
+            "service_data": {"entity_id": self.entity_id},
         }
 
         self.request_id += 1  # Increment the request ID for the next request
@@ -94,7 +103,7 @@ class HomeAssistantWebsocket:
                 todo_list = (
                     response_data.get("result", {})
                     .get("response", {})
-                    .get("todo.shopping_list", {})
+                    .get(self.entity_id, {})
                     .get("items", [])
                 )
                 return [item for item in todo_list]
@@ -121,7 +130,7 @@ class HomeAssistantWebsocket:
             "type": "call_service",
             "domain": "todo",
             "service": "add_item",
-            "service_data": {"entity_id": "todo.shopping_list", "item": item_name},
+            "service_data": {"entity_id": self.entity_id, "item": item_name},
         }
 
         self.request_id += 1  # Increment the request ID for the next request

--- a/alexa_shopping_list_sync/app/constants/entity_id_mapping.py
+++ b/alexa_shopping_list_sync/app/constants/entity_id_mapping.py
@@ -1,0 +1,5 @@
+entity_id_mapping = {
+    "en": "todo.shopping_list",
+    "en-GB": "todo.shopping_list",
+    "de": "todo.einkaufsliste",
+}

